### ISSUE TITLE
[Attribute] Updated database value design change

### ISSUE THREAD

**Issue by peteward** [Sylius/Sylius#3805]
Hi,

I've just updated to the new attribute model.

What's the point of the reworked database design around replacing `value` with 6 different typed versions? I'm trying to understand what problems it solved compared to the problems it creates.

The problems that it has created for our project:
* You can't cleanly query on 'value' now. You have to check which field to use base on the storage_type. This really breaks any kind of efficient shortcut queries that you want to write, and the very nature of the attribute model means that in many scenarios you are quite likely to want to write queries to simplify the work rather than hydrate everything through the application.
* 100,000 products each with a lot of attributes. This has added 5 extra columns of storage - it's not massive at 25 bytes if using MySQL, but if there's 10 attributes per product that's a million rows that this has been added to.

The advantages of doing this to me seem:
* Validation or correct representation at the storage level (this can be achieved and would be preferable through application in my eyes anyway).
* Being able to sort / query at the database level - but this is countered by the above point by spreading values across different columns. Querying / ordering a single attribute is fine IF you know the type and can build the query in advance, otherwise not really.

The solution for us is pretty simple - we'll just make everything 'text' and use 'text_value' - a shame to lose the functionality but it feels flawed with this storage implementation.

Is there some really obvious advantage that I've missed?

**Comment by peteward** [comment]
ping @Zales0123

**Comment by pjedrzejewski** [comment]
Hello Peter! Thanks for opening the issue, let me elaborate on why we believe this is better:

* You can't query on the ``value`` column as easily as before but now you are able to query attribute values using proper types. I strongly believe this is a good trade-off because querying on the attributes in EAV model is ineffective anyway and probably should be one in ElasticSearch indexed documents etc. What do you think?
* Does having these empty columns affects the query performance or is it just about disk space?

---

Advantages in my opinion:

* Using proper storage in the database and being able to query properly based on these types.
* Converting values to proper types in application layer was not good and if we want to introduce more types, we would need to have a lot of ifs etc. now Doctrine handles that for us and we can very easily use all types.
* There is a lot of flexibility for custom attributes now, they can pick proper storage type and use it effectively, can be JSON for example. So you can have an attribute that is an array and store that properly in db. Otherwise our table would have just one "value"  field which would have strings, 1/0, dates with hardcoded format and serialized arrays or json? That does not sound good.
* Like you said, you can still use "text" for all attributes and use proper column. So, while not perfect, you can maintain previous behavior?

I borrowed this idea from Akeneo guys, so would be great if they could share their experience. What are you thoughts on this approach? /cc @nidup, @jjanvier (tl;dr: We use same approach for storing attribute values)

**Comment by peteward** [comment]
These are some good arguments. 

Storage is a bit of an afterthought, it's not a big deal so I'll leave that one.

For me it comes down to efficiency of querying and losing the flexbility to just write a query if you need something. I'm a bit "old school" of liking to have good control and ability at the database layer and I think these changes make that much harder. I think this is more of an EAV model on steroids because the complexity of querying an EAV model just got much harder and slower.

If you're doing everything through the ORM layers then all of your arguments stand and make great sense, but if you want to bypass ORM (which you sometimes do for efficiency purposes), then you're struggling with this model.

Here's a simple method we use for pulling out a keyed array of store attributes that need grouping by category:

```
    /**
     * @param array $store
     *
     * @return array
     */
    public function getAttributesValueByCategory($store)
    {
        $qb = $this->getEntityManager()->createQueryBuilder('attributeValue');

        $qb
            ->select('av.text as value, c.name as categoryName, t.name as name')
            ->from('Reiss\StoreLocatorBundle\Entity\AttributeValue', 'av')
            ->join('av.attribute', 'a')
            ->join('a.category', 'c')
            ->join('a.translations', 't')
            ->where('av.subject = :subject')
            ->setParameter('subject', $store)
        ;

        $values = $qb->getQuery()->getResult();

        $keyedValues = array();
        foreach ($values as $value) {
            if ($value['value']) {
                $keyedValues[$value['categoryName']][$value['name']] = $value['value'];
            }
        }

        return $keyedValues;
    }
```

If we had attribute values in different columns this query becomes very difficult to write and work with. you'll either have some complex custom DQL functions or be unifying values in the application.

We have quite a few methods that retrieve attribute values in bulk like this - including for indexing Elasticsearch (because ORM took 5 minutes compared to 10 seconds on selective array querying). We do it because it's much faster.

Anyway I'll close this because we'll proceed as we are and I know we all have bigger issues to worry about.

**Comment by nidup** [comment]
Hi @peteward and @pjedrzejewski 

Sorry for the late answer, building a Product Information Management (PIM) system based on EAV model for 3 years now, we can do a quick feedback about it :)

EAV storage brings a lot of flexibility to our customers, they can dynamically create new attributes in the system through the UI, no need to ask to a developer, no need of a maintenance time to do so.
For our PIM system, with the use cases we know, as the catalog structure changes often, it's a must have.

There are several way to implement this, using a document oriented storage, using a RDBMS + EAV model, live updating the table schema (which locks the table when updates), etc.
As we want to create a open source PIM usable on a classic LAMP stack, we had this RDBMS "constraint" for the product model from the very beginning.
Our 1.0 has been released with a RDBMS EAV storage (based on Doctrine ORM).
In 1.1, we've introduced a second MongoDB storage for the product data (for implementation with more than 5 Millions of product values).
In 1.1, 1.2, 1.3, 1.4, we introduced several abstractions to manipulate the products in the same way from the developer point of view (whatever the data is stored in RDBMS EAV or Document Oriented storage).

You're right, the drawbacks of this EAV approach is the complexity of querying the data and the poor performance, it's not a bit more complex and slow, it's a huge difference, crazy complex, crazy slow :)
You also encounter new issues, for instance, something as simple as a unique constraint on a table field become a nightmare to implement with a EAV model.

By working a lot, we managed to limit a lot these difficulties for the developers who integrate Akeneo PIM, by introducing lot of internal API classes to ease the product manipulation (more details here http://docs.akeneo.com/latest/cookbook/catalog/product/index.html).

We use this EAV model only for products, we assume that other models (category, user, etc) should be customized during the project development, you don't need to "live change them".

The development cost of a EAV model is very very high, so I would do it only when there is no other choice. This choice depends on a lot of criteria. How many product properties i have? How many product do we have? Do I need to add/change product properties often once the website is online? Do these values change often? Do i need to query these values or is it only for display purpose? etc

In the case of a home-made e-commerce project, depending on the answers to these questions, i would use a different solution, for instance, Elastic Search for full text search, facets, etc

The difficulty is to provide a simple out of the box solution, which will be usable for all the projects using Sylius. The solution working perfectly for every use cases does not exist, so I would advise to provide few simple common facilities and let the developer choose and implement what is the best for his project, maybe a bit more work but more flexibility too :)

Don't hesitate if you want to discuss more about this topic and keep up the good work, Sylius is a real source of inspiration for our core team!

**Comment by peteward** [comment]
Hi @nidup,

Thanks for the feedback, that was a very interesting read!  And I agree that a model with flexibility is the key. 

I think my main criticism of the Sylius model is that the visual form element is tied to the storage type. We were looking to workaround typed storage by having everything in 'text_value' field for easy cross-attribute querying, but it seems that 'checkbox' cannot be used with a text type. 

If Sylius were to allow this then I think the flexibility is there. I was pretty sure that this did work originally when we upgraded to the new model but then I tried again recently and the forms wouldn't render.
